homeassistant/components/restore_state.py

```diff
@@ -0,0 +1,86 @@
+"""
+Support for restoring entity states on startup.
+
+Component that records all events and state changes. Allows other components
+to query this database.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/recorder/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, split_entity_id
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START, ATTR_RESTORED_STATE)
+from homeassistant.components.history import (
+    CONF_DOMAINS, CONF_EXCLUDE, CONF_ENTITIES, CONF_INCLUDE,
+    Filters, get_states, last_recorder_run)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import ConfigType
+import homeassistant.util.dt as dt_util
+
+DOMAIN = 'restore_state'
+
+DEPENDENCES = ['recorder']
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: cv.FILTER_SCHEMA,
+}, extra=vol.ALLOW_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+_ALWAYS_OVERWRITE = (
```

Comment:
We should not have any always overwrite domains. Because now the state machine will no longer reflect the internal state. For example, calling `input_boolean.toggle` service will result in an inconsistent state. Instead, platforms will have to implement this themselves.

To not have to duplicate logic all over the place, I suggest the following: (this might be out of scope for this PR)

 - Ensure that `restore_state` is [one of the first components to be setup](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/loader.py#L195-L198), restoring all the previous entities.
 - Teach the recorder to not save any state_changed event where new_state has `ATTR_RESTORED_STATE`
 - Teach `async_generate_entity_id` to allow generating an entity id that already exists with `ATTR_RESTORED_STATE`.
 - In platforms that want to restore their state, they can retrieve their previous state from the state machine and set data based on that.

Thus far the idea was pretty clear in my head but I need to think more about the last point.

Actually, we could even build this into the entity component / ABC class. We would need a new 'hook' that gets called when an entity is all setup and ready to go, [probably here](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_component.py#L203): `entity.added_to_hass()`. Entity ABC would then implement this and allow restoration, for example in the light class:

```python
class Light(entity):
    def restore_state(self, state, **kwargs):
        """Restore state from previous state.

        Should be overwritten by entities.
        """
        pass

    def added_to_hass(self):
        state = self.hass.states.get(self.entity_id)

        if state is None:
            return

        # Do not just pass the state to restore_state.
        # How properties get written to state machine is an implementation detail.
        self.restore_state(
            state.state,
            state.attributes.get(ATTR_BRIGHTNESS),
            state.attributes.get(ATTR_COLOR_TEMP),
            # etc
        )
```

Using this approach we will create a generic solution with the knowledge of how this works in the components that need to implement it instead of trying to put the knowledge inside `restore_state.py`, which does not scale.

What do you think?